compiler/rustc_hir_analysis/messages.ftl

```diff
@@ -463,24 +463,20 @@ hir_analysis_start_not_target_feature = `#[start]` function is not allowed to ha
 hir_analysis_start_not_track_caller = `#[start]` function is not allowed to be `#[track_caller]`
     .label = `#[start]` function is not allowed to be `#[track_caller]`
 
-hir_analysis_static_mut_ref = creating a {$shared} reference to a mutable static
-    .label = {$shared} reference to mutable static
-    .note = {$shared ->
-        [shared] this shared reference has lifetime `'static`, but if the static ever gets mutated, or a mutable reference is created, then any further use of this shared reference is Undefined Behavior
-        *[mutable] this mutable reference has lifetime `'static`, but if the static gets accessed (read or written) by any other means, or any other reference is created, then any further use of this mutable reference is Undefined Behavior
-    }
+hir_analysis_static_mut_ref = creating a {$shared_label}reference to a mutable static
+    .label = {$shared_label}reference to mutable static
+    .shared_note = shared references to mutable statics are dangerous since if there's any kind of mutation of, or mutable reference created for, that static while the reference lives, that's undefined behavior
+    .mut_note = mutable references to mutable statics are dangerous since if there's any other pointer used or reference created for that static while the reference lives, that's undefined behavior
     .suggestion = use `addr_of!` instead to create a raw pointer
     .suggestion_mut = use `addr_of_mut!` instead to create a raw pointer
 
-hir_analysis_static_mut_refs_lint = creating a {$shared} reference to mutable static is discouraged
-    .label = {$shared} reference to mutable static
+hir_analysis_static_mut_refs_lint = creating a {$shared_label}reference to mutable static is discouraged
+    .label = {$shared_label}reference to mutable static
     .suggestion = use `addr_of!` instead to create a raw pointer
     .suggestion_mut = use `addr_of_mut!` instead to create a raw pointer
     .note = this will be a hard error in the 2024 edition
-    .why_note = {$shared ->
-        [shared] this shared reference has lifetime `'static`, but if the static ever gets mutated, or a mutable reference is created, then any further use of this shared reference is Undefined Behavior
-        *[mutable] this mutable reference has lifetime `'static`, but if the static gets accessed (read or written) by any other means, or any other reference is created, then any further use of this mutable reference is Undefined Behavior
-    }
+    .shared_note = shared references to mutable statics are dangerous since if there's any kind of mutation of, or mutable reference created for, that static while the reference lives, that's undefined behavior
```

Comment:
```suggestion
    .shared_note = shared references to mutable statics are dangerous. If this static is mutated or a mutable reference to it exists, creating another reference is undefined behavior
```

I got a bit confused reading this note in the error message. Also it's a bit too long IMO. First, I would split it up into two sentences. Then I would word the second part a bit differently.

Looking at my own suggestion and to remove the punctuation from it, this should maybe be split into 2 notes. Or you can leave the sentence about `dangerous` out completely. The note about UB at the end already scream danger.

Reply:
You're not wrong about it being dense.  The trouble is that it's very carefully worded, and so it's easy in trying to change it for better readability to make it slightly wrong (as the version you suggest unfortunately is).  That said, it probably is possible to do a bit better here.  E.g.:

```suggestion
    .shared_note = shared references to mutable statics are dangerous; it's undefined behavior if the static is mutated or if a mutable reference is created for it while the shared reference lives
```

Reply:
Thanks for improving it. So it is not UB, if the mutable reference is created first and then the shared reference is created later, like I wrote in my suggestion? Or is it UB, if the mutable reference and the shared reference live at the same time, no matter the creation order? In the later case, I think your suggestion (and the original) is also not fully accurate.

Reply:
The case you mention is covered by the language for mutable references:

> mutable references to mutable statics are dangerous; it's undefined behavior if any other pointer to the static is used or if any other reference is created for the static while the mutable reference lives

Reply:
(The reason your version was incorrect wasn't this.  It's that you had said "if A or B, and if C happens, then UB", when "if A or B then UB" would be more correct, and that it dropped the condition of liveness from the predicate.)

Reply:
The unfortunate bit about trying to describe this is that we're stating the somewhat obvious rules (anywhere in Rust, it's always UB to create a shared reference that aliases a mutable one) but not really getting at *why* this in particular is dangerous.  The mutable statics are dangerous because of how uniquely easy these statics make it to violate these rules.  I don't have any immediate ideas about how to concisely encode that in an error message, though.